### PR TITLE
HDDS-8127. Exclude deleted containers from Recon container count

### PR DESCRIPTION
**Pull request by ArafatKhan2198**
## What changes were proposed in this pull request?

The deleted container count is being fetched to subtract it from the total container count. This is because the count of deleted containers is not relevant for the current state of the cluster, and including it in the total container count would provide misleading information about the number of available containers. Therefore, it is necessary to subtract the deleted container count from the total container count to obtain the actual number of available containers. This adjusted container count is then used to construct the ClusterStateResponse that is returned by the getClusterState() method.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8127

## How was this patch tested?

Tested it manually by deleting containers from the cluster

### CODE REVIEW

**Comment by ArafatKhan2198** [comment]
@devmadhuu @sadanand48 @dombizita can you please take a look 
Thanks !

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java [hunk 1]
```diff
@@ -134,7 +136,8 @@ public Response getClusterState() {
       totalKeys += fileRecord.getValue();
     }
     builder.setKeys(totalKeys);
-
+    // Subtract deleted containers from total containers.
+    containers = containers - deletedContainersCount;
```
Comment (sadanand48) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java:140]: A unit test would be good to have for this.

Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java:140]: Will do just wanted to confirm if the approach was correct. !

Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java:140]: I think we should also expose in UI this deletedContainer count separately a part from maths computation and updating total container count.

Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java]: Got it I have exposed the `deletedContainersCount` to the response as well !!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/ClusterStateResponse.java [hunk 2]
```diff
@@ -66,6 +66,12 @@ public final class ClusterStateResponse {
   @JsonProperty("openContainers")
   private int openContainers;
 
+  /**
+   * Total count of deleted containers in the cluster.
+   */
+  @JsonProperty("deletedContainers")
+  private int deletedContainers;
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/ClusterStateResponse.java:73]: I have added a class called as `ContainerStats` that will map the count of various container state counts.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java (hunk 1 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java]: @sadanand48 I have added the UT's

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java [hunk 3]
```diff
@@ -134,15 +150,19 @@ public Response getClusterState() {
       totalKeys += fileRecord.getValue();
     }
     builder.setKeys(totalKeys);
-
+    // Subtract deleted containers from total containers.
+    containerStateCounts.setTotalContainerCount(
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java:154]: @ArafatKhan2198 , pls call this setter method for total container count once at last for final value. You can remove the call from line no 95 above.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/ClusterStateResponse.java [hunk 4]
```diff
@@ -104,6 +110,7 @@ private ClusterStateResponse(Builder b) {
     this.containers = b.containers;
     this.missingContainers = b.missingContainers;
     this.openContainers = b.openContainers;
+    this.deletedContainers = b.deletedContainers;
```
Comment (sumitagrawl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/ClusterStateResponse.java:113]: currently we have a TestTotalOpenContainerCount and TestTotalDeletedContainerCount test classes for each different state counts we can consolidate all of them into a single test class that calculates the container count of different states.
We could call that class TestContainerStateCounts

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java (hunk 3 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java:175]: Thanks for the suggestion @devmadhuu will implement it !!

**Review by sumitagrawl — APPROVED** [review]
@ArafatKhan2198 Thanks working on this, LGTM +1

**Pull request merged**